src/de/impl_tuples.rs

```diff
@@ -126,16 +250,607 @@ where
+        Ok((
+            A::decode(decoder)?,
+            B::decode(decoder)?,
+            C::decode(decoder)?,
+            D::decode(decoder)?,
+            E::decode(decoder)?,
+            F::decode(decoder)?,
+            G::decode(decoder)?,
+            H::decode(decoder)?,
+            I::decode(decoder)?,
+            J::decode(decoder)?,
+            K::decode(decoder)?,
+            L::decode(decoder)?,
+            M::decode(decoder)?,
+            N::decode(decoder)?,
+            O::decode(decoder)?,
+        ))
+    }
+}
+impl<'de, A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P> BorrowDecode<'de>
+    for (A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P)
+where
+    A: BorrowDecode<'de>,
+    B: BorrowDecode<'de>,
+    C: BorrowDecode<'de>,
+    D: BorrowDecode<'de>,
+    E: BorrowDecode<'de>,
+    F: BorrowDecode<'de>,
+    G: BorrowDecode<'de>,
+    H: BorrowDecode<'de>,
+    I: BorrowDecode<'de>,
+    J: BorrowDecode<'de>,
+    K: BorrowDecode<'de>,
+    L: BorrowDecode<'de>,
+    M: BorrowDecode<'de>,
+    N: BorrowDecode<'de>,
+    O: BorrowDecode<'de>,
+    P: BorrowDecode<'de>,
+{
+    fn borrow_decode<BD: BorrowDecoder<'de>>(decoder: &mut BD) -> Result<Self, DecodeError> {
+        Ok((
+            A::borrow_decode(decoder)?,
+            B::borrow_decode(decoder)?,
+            C::borrow_decode(decoder)?,
+            D::borrow_decode(decoder)?,
+            E::borrow_decode(decoder)?,
+            F::borrow_decode(decoder)?,
+            G::borrow_decode(decoder)?,
+            H::borrow_decode(decoder)?,
+            I::borrow_decode(decoder)?,
+            J::borrow_decode(decoder)?,
+            K::borrow_decode(decoder)?,
+            L::borrow_decode(decoder)?,
+            M::borrow_decode(decoder)?,
+            N::borrow_decode(decoder)?,
+            O::borrow_decode(decoder)?,
+            P::borrow_decode(decoder)?,
+        ))
+    }
+}
+impl<A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P> Decode
+    for (A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P)
+where
+    A: Decode,
+    B: Decode,
+    C: Decode,
+    D: Decode,
+    E: Decode,
+    F: Decode,
+    G: Decode,
+    H: Decode,
+    I: Decode,
+    J: Decode,
+    K: Decode,
+    L: Decode,
+    M: Decode,
+    N: Decode,
+    O: Decode,
+    P: Decode,
+{
+    fn decode<DE: Decoder>(decoder: &mut DE) -> Result<Self, DecodeError> {
+        Ok((
+            A::decode(decoder)?,
+            B::decode(decoder)?,
+            C::decode(decoder)?,
+            D::decode(decoder)?,
+            E::decode(decoder)?,
+            F::decode(decoder)?,
+            G::decode(decoder)?,
+            H::decode(decoder)?,
+            I::decode(decoder)?,
+            J::decode(decoder)?,
+            K::decode(decoder)?,
+            L::decode(decoder)?,
+            M::decode(decoder)?,
+            N::decode(decoder)?,
+            O::decode(decoder)?,
+            P::decode(decoder)?,
+        ))
+    }
+}
+
+// For re-generating the above code
```

Comment:
Why don't we just keep the macro code instead of having the manual expansion?

Reply:
I thought the macro would be slower than the exported code. Testing this seems to be partly true, but not enough to matter:

Build times were ~3.48 seconds with the exported code, and ~3.52 seconds with the macro.